pkg/rpc/peer_test.go

```diff
@@ -0,0 +1,204 @@
+	k := peerKey{TargetAddr: targetAddr, NodeID: nodeID, Class: class}
+	return m.kh(k, healthy)
+}
+
+func Test_hasSiblingConn(t *testing.T) {
+	const (
+		t1  = "1.1.1.1:1111"
+		t2  = "2.2.2.2:2222"
+		t3  = "3.3.3.3:3333"
+		n1  = roachpb.NodeID(1)
+		n2  = roachpb.NodeID(2)
+		n3  = roachpb.NodeID(3)
+		def = DefaultClass
+		sys = SystemClass
+	)
+
+	defSelf := func() peerKey {
+		return peerKey{
+			TargetAddr: t1,
+			NodeID:     n1,
+			Class:      def,
+		}
+	}
+
+	defZeroSelf := func() peerKey {
+		return peerKey{
+			TargetAddr: t1,
+			Class:      def,
+		}
+	}
+
+	tests := []struct {
+		name        string
+		m           pmb
+		self        peerKey
+		wantHealthy bool
+		wantOk      bool
+	}{
+		// In the empty map, defSelf has no sibling.
+		{
+			name: "empty",
+		},
+
+		// First, we test various combinations where our peerKey has a NodeID.
+		// We'll test the NodeID==0 case below.
+
+		// In a map that contains only itself, defSelf has no sibling.
+		{
+			name: "single",
+			m:    (&pmb{}).k(defSelf()),
+		},
+		// In a map that contains defSelf and another healthy but unrelated peer,
+		// defSelf finds no sibling.
+		{
+			name: "unrelated",
+			m:    (&pmb{}).k(defSelf()).p(n2, t2, sys, true /* healthy */),
+		},
+		// Same as previous, but with a curveball: n2 also hails from defSelf's
+		// address. (In other words, n2 must have restarted with a listener
+		// formerly used by n1, or vice versa). Not a sibling - we're looking
+		// for the same NodeID on a different TargetAddr.
+		{
+			name: "unrelated-switcheroo",
+			m:    (&pmb{}).k(defSelf()).p(n2, t1, def, true /* healthy */),
+		},
+		// Unrelated peer that has a zero NodeID.
+		{
+			name: "unrelated-switcheroo",
+			m:    (&pmb{}).k(defSelf()).p(0, t3, def, true /* healthy */),
+		},
+		// If a healthy sibling is present, it is found.
+		{
+			name:   "related-healthy",
+			m:      (&pmb{}).k(defSelf()).p(n1, t2, def, true /* healthy */),
+			wantOk: true, wantHealthy: true,
+		},
+		// Like last, but sibling has a different connection class - outcome is the
+		// same.
+		{
+			name:   "related-healthy-sys",
+			m:      (&pmb{}).k(defSelf()).p(n1, t2, sys, true /* healthy */),
+			wantOk: true, wantHealthy: true,
+		},
+		// If an unhealthy sibling is present, it is found.
+		{
+			name:   "related-unhealthy",
+			m:      (&pmb{}).k(defSelf()).p(n1, t2, def, false /* healthy */),
+			wantOk: true, wantHealthy: false,
+		},
+		// Ditto with other class.
+		{
+			name:   "related-unhealthy-sys",
+			m:      (&pmb{}).k(defSelf()).p(n1, t2, sys, false /* healthy */),
+			wantOk: true, wantHealthy: false,
+		},
+
+		// Now we test the NodeID == 0 case.
+
+		// Peer is not sibling of self.
+		{
+			name: "zero-single",
+			m:    (&pmb{}).k(defZeroSelf()),
+			self: defZeroSelf(),
+		},
+		// Peer is not sibling of other peers that have non-matching target addrs.
+		{
+			name: "zero-unrelated",
+			m: (&pmb{}).
```

Comment:
Fluent map builder is fancy, but isn't `pmb(defSelf(), p(n,t,c,h))` as a func with varargs more readable because less () and {}?
I like simple, but YMMV so up to you.

Reply:
You're right, but it's a little awkward because the mock `peer` I construct here doesn't know its own `connKey`. I still think your suggestion is directionally right, but I'm going to leave as is.